File: bindings/dotnet/examples/CompileOnceEvaluateMany.cs
// compile-once-evaluate-many: compile a rule once, evaluate it against
// ~100k payloads, and print the last result plus a rough per-evaluation cost.
//
// Run from bindings/dotnet/:
//   dotnet run --project examples -- compile-once

using System.Diagnostics;

namespace Goplasmatic.Datalogic.Examples;

internal static class CompileOnceEvaluateMany
{
    private const int Iterations = 100_000;

    internal static void Run()
    {
        using var engine = new Engine();
        using var rule = engine.Compile(
            """{"*": [{"var": "price"}, {"-": [1, {"var": "discount"}]}]}""");

        var last = "";
        var stopwatch = Stopwatch.StartNew();
        for (var i = 0; i < Iterations; i++)
        {
            last = rule.Evaluate($$"""{"price": {{100 + i % 100}}, "discount": 0.2}""");
        }
        stopwatch.Stop();

        Console.WriteLine($"last result: {last}");
        Console.WriteLine(
            $"{Iterations} evaluations, ~{stopwatch.Elapsed.TotalNanoseconds / Iterations:F0} ns/op");
    }
}

File: bindings/php/examples/compile-once-evaluate-many.php
<?php

declare(strict_types=1);

// compile-once-evaluate-many: compile a rule once, evaluate it against
// ~100k payloads, and print the last result plus a rough per-evaluation cost.
//
// Run from bindings/php/ (needs `composer install` once, plus the C ABI cdylib:
// `cargo build --release --manifest-path ../c/Cargo.toml`):
//   php examples/compile-once-evaluate-many.php

use Goplasmatic\Datalogic\Engine;

require __DIR__ . '/../vendor/autoload.php';

const ITERATIONS = 100_000;

$engine = new Engine();
$rule = $engine->compile('{"*": [{"var": "price"}, {"-": [1, {"var": "discount"}]}]}');

$last = null;
$start = hrtime(true);
for ($i = 0; $i < ITERATIONS; $i++) {
    $last = $rule->evaluate(json_encode(['price' => 100 + $i % 100, 'discount' => 0.2]));
}
$elapsedNs = hrtime(true) - $start;

echo "last result: {$last}", PHP_EOL;
echo sprintf('%d evaluations, ~%d ns/op', ITERATIONS, intdiv($elapsedNs, ITERATIONS)), PHP_EOL;
